File: card-lang/src/calculator/lexer/tokens/identifier.rs
use std::collections::HashMap;

use crate::types::Lexer::Lexer;
use crate::types::Lexer::Tokenizer;
#[path = "./operator.rs"] mod op;
#[path = "./digit.rs"] mod digit;
#[path = "./whitespace.rs"] mod ws;

pub struct IdentifierTokenizer{}

impl Tokenizer for IdentifierTokenizer {
  fn token_type(&self)->String{
    return "identifier".to_string();
  }
  fn matchesChar(&self, input: char)->bool{
    return matchesChar(input)
  }
  fn handleChar(
    &self, c: char, lexer: Lexer
  )->Result<(), String>{
    return handleChar(c, lexer);
  }
}

pub fn matchesChar(input: char) -> bool{
  return !op::matchesChar(input) && !digit::matchesChar(input) && !ws::matchesChar(input);
}

fn handleChar(initial_char: char, mut lexer: Lexer) -> Result<(), String> {
  let mut l = lexer;
  let mut c = initial_char;
  let mut identity = "".to_string();
  identity.push(c);
  loop {
    (l, c) = l.advance();
    if !matchesChar(c) {
      break;
    }
    identity.push(c);
  }
  l = l.addToken(HashMap::from([("name".to_string(), identity.to_owned())]));
  l.advance();
  return Ok(());
}

File: card-lang-typescript/src/calculator/lexer/tokens/identifier.ts
import { LexerController, Tokenizer } from "../../../types/Lexer";

import { matchesChar as opMatchesChar } from "./operator";
import { matchesChar as digMatchesChar } from "./number";
import { matchesChar as wsMatchesChar } from "./whitespace";

export class IdentifierTokenizer implements Tokenizer {
  token_type = "identifier";

  matchesChar(input: string){
    return matchesChar(input)
  }
  handleChar(c: string, lexer: LexerController){
    return handleChar(c, lexer);
  }

}

function matchesChar(input: string){
  return !opMatchesChar(input) && !digMatchesChar(input) && !wsMatchesChar(input);
}

function handleChar(initial_char: string, lexer: LexerController) {
  var c: string | void;
  let identity = initial_char;
  while(true) {
    c = lexer.advance();
    if(typeof c == "undefined" || !matchesChar(c)){ break; }
    identity += c;
  }
  lexer.addToken({ name: identity });
  lexer.advance();
}